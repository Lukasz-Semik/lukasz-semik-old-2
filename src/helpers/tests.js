import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';

const localVue = createLocalVue();

localVue.use(Vuex);

export const generateTestVuex = () => ({
  localVue,
  store: new Vuex.Store({}),
});

export const findByTestAttr = (node, attrName) => node.find(`[data-test="${attrName}"]`);
export const findAllByTestAttr = (node, attrName) => node.findAll(`[data-test="${attrName}"]`);
