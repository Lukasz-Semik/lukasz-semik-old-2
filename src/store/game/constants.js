export const gameStates = {
  pristine: 'pristine',
  intro: 'intro',
  counting: 'counting',
  running: 'running',
};
