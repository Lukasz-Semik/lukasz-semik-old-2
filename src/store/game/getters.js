export const names = {
  gameState: 'gameState',
};

export default {
  [names.gameState]: state => state.gameState,
};
