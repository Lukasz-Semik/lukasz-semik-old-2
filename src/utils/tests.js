import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';

const localVue = createLocalVue();

localVue.use(Vuex);

export const generateTestGameVuex = ({
  getters = {
    isGamePaused: () => false,
  },
  mutations = {},
  actions = {},
  state = {},
} = {}) => ({
  localVue,
  store: new Vuex.Store({
    getters,
    mutations,
    actions,
    state,
  }),
});

export const findByTestAttr = (node, attrName) => node.find(`[data-test="${attrName}"]`);
export const findAllByTestAttr = (node, attrName) => node.findAll(`[data-test="${attrName}"]`);
