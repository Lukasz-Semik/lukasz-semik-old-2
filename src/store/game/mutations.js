import { game, initialState } from './constants';

export const names = {
  setGameIntroState: 'setGameIntroState',
  setGameCountingState: 'setGameCountingState',
  setGameRunningState: 'setGameRunningState',
  substractHealthPoints: 'substractHealthPoints',
  // TODO: make more valid version
  updateGameStatistics: 'updateGameStatistics',
  setIsGamePaused: 'setIsGamePaused',
  restartGame: 'restartGame',
};

export default {
  [names.setGameIntroState]: state => {
    state.gameState = game.hasIntroState;
  },
  [names.setGameCountingState]: state => {
    state.gameState = game.hasCountingState;
  },
  [names.restartGame]: (state, newGameState = game.hasCountingState) => {
    const newState = {
      ...initialState,
      gameState: newGameState,
    };

    Object.assign(state, newState);
  },
  [names.setGameRunningState]: state => {
    state.gameState = game.hasRunningState;
  },
  [names.setIsGamePaused]: (state, isGamePaused) => {
    state.isGamePaused = isGamePaused;
  },
  [names.substractHealthPoints]: state => {
    if (state.healthPoints <= 5) {
      state.gameState = game.hasOverState;
      state.isGamePaused = true;
    }

    // state.healthPoints--;
    state.healthPoints -= 5;
  },
  [names.updateGameStatistics]: state => {
    const newGameScore = state.gameScore + 1;

    if (newGameScore % 5 === 0) {
      state.gameGold++;
    }

    state.gameScore = newGameScore;
  },
};
