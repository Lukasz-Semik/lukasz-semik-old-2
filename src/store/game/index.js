import getters, { names as gettersNames } from './getters';
import mutations, { names as mutationsNames } from './mutations';

const state = {
  gameState: 'pristine',
};

export default {
  state,
  getters,
  mutations,
};

export const game = {
  ...gettersNames,
  ...mutationsNames,
};
