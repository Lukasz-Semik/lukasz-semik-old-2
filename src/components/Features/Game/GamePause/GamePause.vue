<template>
  <modal-element :name="modalName" :title="$t('underwater.pause')">
    <menu-group
      @handleResumeClick="onResumeClick"
      @handleRestartClick="onRestartClick"
      @handleIntroClick="onIntroClick"
      @handleHomeClick="onHomeClick"
    />
  </modal-element>
</template>

<script>
import { mapMutations } from 'vuex';

import { game } from '@/store/game';
import { ModalElement } from '@/components/Elements';

import MenuGroup from './MenuGroup/MenuGroup';

const { setIsGamePaused, restartGame } = game;

export default {
  name: 'GamePause',
  components: {
    ModalElement,
    MenuGroup,
  },
  created() {
    if (window) {
      window.onblur = () => {
        this.setIsGamePaused(true);
        this.$modal.show(this.modalName);
      };
    }
  },
  computed: {
    modalName() {
      return 'pause-modal';
    },
  },
  methods: {
    ...mapMutations({ setIsGamePaused, restartGame }),
    onItemClick(stateChanger) {
      stateChanger();
      this.$modal.hide(this.modalName);
    },
    onResumeClick() {
      this.onItemClick(() => this.setIsGamePaused(false));
    },
    onRestartClick() {
      this.onItemClick(() => this.restartGame());
    },
    onIntroClick() {
      this.onItemClick(() => this.restartGame(game.hasIntroState));
    },
    onHomeClick() {
      this.onItemClick(() => this.restartGame(game.hasPristineState));
    },
  },
};
</script>
