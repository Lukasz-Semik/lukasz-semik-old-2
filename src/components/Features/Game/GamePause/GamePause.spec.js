import { shallowMount } from '@vue/test-utils';

import { generateTestGameVuex } from '@/utils/tests';
import { game } from '@/store/game';

import GamePause from './GamePause';

describe('<GamePause />', () => {
  it('should render properly', () => {
    expect(
      shallowMount(GamePause, {
        mocks: {
          $t: key => key,
        },
        ...generateTestGameVuex({
          getters: {
            gameScore: () => 1,
            gameState: () => game.hasRunningState,
          },
        }),
      })
    ).toMatchSnapshot();
  });
});
