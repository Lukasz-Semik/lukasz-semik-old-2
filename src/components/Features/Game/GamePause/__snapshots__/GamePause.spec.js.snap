// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GamePause /> should render properly 1`] = `
<modal-element-stub name="pause-modal" title="underwater.pause">
  <menu-group-stub gamestate="hasRunningState" gamescore="1"></menu-group-stub>
</modal-element-stub>
`;
